FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp

```diff
@@ -532,111 +719,241 @@ void Arm64Emitter::FillStaticRegs(bool FPRs, uint32_t GPRFillMask, uint32_t FPRF
+  for (; i < Regs.size(); i += 2) {
+    const auto Reg1 = Regs[i];
+    const auto Reg2 = Regs[i + 1];
     ldp<ARMEmitter::IndexType::POST>(Reg1.X(), Reg2.X(), ARMEmitter::Reg::rsp, 16);
   }
+}
+
+void Arm64Emitter::PushDynamicRegsAndLR(FEXCore::ARMEmitter::Register TmpReg) {
+  const auto CanUseSVE = EmitterCTX->HostFeatures.SupportsAVX;
+  const auto GPRSize = (ConfiguredDynamicRegisterBase.size() + 1) * Core::CPUState::GPR_REG_SIZE;
+  const auto FPRRegSize = CanUseSVE ? Core::CPUState::XMM_AVX_REG_SIZE
+                                    : Core::CPUState::XMM_SSE_REG_SIZE;
+  const auto FPRSize = GeneralFPRegisters.size() * FPRRegSize;
+  const uint64_t SPOffset = AlignUp(GPRSize + FPRSize, 16);
+
+  sub(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::rsp, ARMEmitter::Reg::rsp, SPOffset);
+
+  // rsp capable move
+  add(ARMEmitter::Size::i64Bit, TmpReg, ARMEmitter::Reg::rsp, 0);
+
+  LOGMAN_THROW_A_FMT(GeneralFPRegisters.size() % 2 == 0, "Needs to have multiple of 2 FPRs for RA");
+
+  // Push the vector registers
+  PushVectorRegisters(TmpReg, CanUseSVE, GeneralFPRegisters);
+
+  // Push the general registers.
+  PushGeneralRegisters(TmpReg, ConfiguredDynamicRegisterBase);
+
+  str(ARMEmitter::XReg::lr, TmpReg, 0);
+}
+
+void Arm64Emitter::PopDynamicRegsAndLR() {
+  const auto CanUseSVE = EmitterCTX->HostFeatures.SupportsAVX;
+
+  // Pop vectors first
+  PopVectorRegisters(CanUseSVE, GeneralFPRegisters);
+
+  // Pop GPRs second
+  PopGeneralRegisters(ConfiguredDynamicRegisterBase);
 
   ldr<ARMEmitter::IndexType::POST>(ARMEmitter::XReg::lr, ARMEmitter::Reg::rsp, 16);
 }
 
+void Arm64Emitter::SpillForPreserveAllABICall(FEXCore::ARMEmitter::Register TmpReg, bool FPRs) {
+  const auto CanUseSVE = EmitterCTX->HostFeatures.SupportsAVX;
+  const auto FPRRegSize = CanUseSVE ? Core::CPUState::XMM_AVX_REG_SIZE
+                                    : Core::CPUState::XMM_SSE_REG_SIZE;
+
+  if (EmitterCTX->Config.Is64BitMode()) {
+    std::span<const FEXCore::ARMEmitter::Register> DynamicGPRs = x64::PreserveAll_Dynamic;
+    std::span<const FEXCore::ARMEmitter::VRegister> DynamicFPRs = x64::PreserveAll_DynamicFPR;
+
+    if (CanUseSVE) {
+      DynamicFPRs = x64::PreserveAll_DynamicFPRSVE;
+    }
+
+    const auto GPRSize = AlignUp(DynamicGPRs.size(), 2) * Core::CPUState::GPR_REG_SIZE;
+    const auto FPRSize = DynamicFPRs.size() * FPRRegSize;
+    const uint64_t SPOffset = AlignUp(GPRSize + FPRSize, 16);
+
+    // Spill static registers first.
+    SpillStaticRegs(TmpReg, true, x64::PreserveAll_SRAMask, CanUseSVE ? x64::PreserveAll_SRAFPRSVEMask : x64::PreserveAll_SRAFPRMask);
+
+    sub(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::rsp, ARMEmitter::Reg::rsp, SPOffset);
+
+    // rsp capable move
+    add(ARMEmitter::Size::i64Bit, TmpReg, ARMEmitter::Reg::rsp, 0);
+
+    // Push the vector registers
+    PushVectorRegisters(TmpReg, CanUseSVE, DynamicFPRs);
+
+    // Push the general registers.
+    PushGeneralRegisters(TmpReg, DynamicGPRs);
+  }
+  else {
```

Comment:
surely we can/should dedupe the two sides of the if?

Reply:
Deduplicated